File: backend/src/main/java/com/construction/material/service/LicenseService.java
package com.construction.material.service;

import com.construction.material.dto.request.LicensePlanRequest;
import com.construction.material.dto.request.GenerateLicenseRequest;
import com.construction.material.dto.response.LicenseGenerationResponse;
import com.construction.material.dto.response.LicenseMineResponse;
import com.construction.material.dto.response.LicensePlanResponse;
import com.construction.material.dto.response.LicenseResponse;
import com.construction.material.entity.Company;
import com.construction.material.entity.License;
import com.construction.material.entity.LicensePlan;

import java.util.List;

public interface LicenseService {

    // --- Plans (catalog configuration) ---
    List<LicensePlanResponse> getPlans();
    LicensePlanResponse updatePlan(LicensePlan.LicenseType type, LicensePlanRequest request);

    // --- Super Admin: grant lifecycle ---
    LicenseGenerationResponse generateLicense(Long companyId, GenerateLicenseRequest request, String actorUsername);
    LicenseResponse revoke(Long companyId, String actorUsername);
    LicenseResponse suspend(Long companyId);
    LicenseResponse reactivate(Long companyId);
    LicenseResponse getLatest(Long companyId);
    List<LicenseResponse> getHistory(Long companyId);

    // --- Enforcement (used by LicenseEnforcementFilter and ModuleAccessGuard) ---
    LicenseEnforcementStatus resolveEnforcementStatus(Long companyId);

    // --- Self-service ---
    /** Issues an immediately-ACTIVE TRIAL license with no key, for self-registered companies. */
    License issueTrialLicense(Company company);
    LicenseResponse activateByKey(Long companyId, String key);
    LicenseMineResponse getMine(Long companyId);

    // --- Platform settings ---
    boolean isSelfRegistrationEnabled();
    void setSelfRegistrationEnabled(boolean enabled);
}

File: frontend/src/app/core/services/license.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import {
  GenerateLicenseRequest,
  License,
  LicenseGenerationResponse,
  LicensePlan,
  LicensePlanUpdateRequest,
  LicenseType
} from '../models/license.model';
import { ApiService } from './api.service';

/** Super Admin only: license plan configuration and per-company license lifecycle. */
@Injectable({
  providedIn: 'root'
})
export class LicenseService {
  private apiUrl = '/admin';

  constructor(private api: ApiService) {}

  getPlans(): Observable<LicensePlan[]> {
    return this.api.get<LicensePlan[]>(`${this.apiUrl}/license-plans`);
  }

  updatePlan(type: LicenseType, request: LicensePlanUpdateRequest): Observable<LicensePlan> {
    return this.api.put<LicensePlan>(`${this.apiUrl}/license-plans/${type}`, request);
  }

  generateLicense(companyId: number, request: GenerateLicenseRequest): Observable<LicenseGenerationResponse> {
    return this.api.post<LicenseGenerationResponse>(`${this.apiUrl}/companies/${companyId}/license/generate`, request);
  }

  revokeLicense(companyId: number): Observable<License> {
    return this.api.post<License>(`${this.apiUrl}/companies/${companyId}/license/revoke`, {});
  }

  suspendLicense(companyId: number): Observable<License> {
    return this.api.post<License>(`${this.apiUrl}/companies/${companyId}/license/suspend`, {});
  }

  reactivateLicense(companyId: number): Observable<License> {
    return this.api.post<License>(`${this.apiUrl}/companies/${companyId}/license/reactivate`, {});
  }

  getCompanyLicenseHistory(companyId: number): Observable<License[]> {
    return this.api.get<License[]>(`${this.apiUrl}/companies/${companyId}/license/history`);
  }

  getSelfRegistrationEnabled(): Observable<{ enabled: boolean }> {
    return this.api.get<{ enabled: boolean }>(`${this.apiUrl}/settings/self-registration`);
  }

  setSelfRegistrationEnabled(enabled: boolean): Observable<{ message: string }> {
    return this.api.put<{ message: string }>(`${this.apiUrl}/settings/self-registration?enabled=${enabled}`, {});
  }
}
